### PR TITLE
fix: patch codemod typography-sizes for picasso v17

### PR DESCRIPTION
**Pull request by OndrejTuma**
[FX-NNNN]

### Description

patched a bug that calculated wrong size in nested Typography components

### How to test

- FIXME: Add the steps describing how to verify your changes

### Screenshots

| Before.                                 | After.                                  |
| --------------------------------------- | --------------------------------------- |
| Insert screenshots or screen recordings | Insert screenshots or screen recordings |

### Review

- [ ] Read [CONTRIBUTING.md](https://github.com/toptal/picasso/blob/master/CONTRIBUTING.md) and [Component API principles](https://github.com/toptal/picasso/blob/master/docs/contribution/component-api.md)
- [ ] Annotate all `props` in component with documentation
- [ ] Create `examples` for component
- [ ] Ensure that deployed demo has expected results and good examples
- [ ] Ensure that tests pass by running `yarn test`
- [ ] Ensure that visuals tests pass by running `yarn test:visual`. If not - check the documentation [how to fix visual tests](https://github.com/toptal/picasso/blob/master/docs/contribution/visual-testing.md#fixing-broken-visual-tests-inside-a-pr)
- [ ] Ensure the changed/created components have not caused accessibility issues. [How to use accessibility plugin in storybook](https://github.com/toptal/picasso/blob/master/docs/contribution/accessibility.md).

<details>
<summary>PR commands</summary>
<br />

List of available commands:

- `@toptal-bot run all` - Run whole pipeline
- `@toptal-bot run build` - Check build
- `@toptal-bot run visual` - Run visual tests
- `@toptal-bot run deploy:documentation` - Deploy documentation
- `@toptal-bot run package:alpha-release` - Release alpha version

</details>

<details>
<summary>PR Review Guidelines</summary>
<br />

#### When to approve? ✅

**You are OK** with merging this PR and
1. You have no extra requests.
2. You have optional requests.
   1. Add `nit: ` to your comment. (ex. `nit: I'd rename this variable from makeCircle to getCircle`)

#### When to request changes? ❌

**You are not OK** with merging this PR because
1. Something is broken after the changes.
2. Acceptance criteria is not reached.
3. Code is dirty.

#### When to comment (neither ✅ nor ❌)

**You want your comments to be addressed** before merging this PR in cases like:
1. There are leftovers like unnecessary logs, comments, etc.
2. You have an opinionated comment regarding the code that requires a discussion.
3. You have questions.

#### How to handle the comments?

1. An owner of a comment is the only one who can resolve it.
2. An owner of a comment must resolve it when it's addressed.
3. A PR owner must reply with ✅ when a comment is addressed.

</details>

### CODE REVIEW

File: .changeset/thirty-bottles-complain.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+'@toptal/picasso-codemod': patch
+---
+
+patched a bug that calculated wrong size in nested Typography components
```
Comment (teimurjan) [inline .changeset/thirty-bottles-complain.md:5]:
```suggestion
Fixed invalid size calculation in nested Typography components
```

File: packages/picasso-codemod/src/v17.0.0/typography-sizes/typography-sizes.ts [hunk 2]
```diff
@@ -10,8 +13,14 @@ const getNewSize = (oldSize: string): string => {
       return oldSize
   }
 }
-
-let affectedComponents = ['Typography', 'TypographyOverflow', 'Amount']
+// exclude components with `variant` prop other than body
+const shouldBeExcluded = (attributes: JSXAttribute[]) =>
+  attributes.find(
```
Comment (teimurjan) [inline packages/picasso-codemod/src/v17.0.0/typography-sizes/typography-sizes.ts:18]: `find` returns an item matching the condition. `some` will return you `true` or `false`.

```suggestion
  attributes.some(
```

File: packages/picasso-codemod/src/v17.0.0/typography-sizes/typography-sizes.ts [hunk 3]
```diff
@@ -10,8 +13,14 @@ const getNewSize = (oldSize: string): string => {
       return oldSize
   }
 }
-
-let affectedComponents = ['Typography', 'TypographyOverflow', 'Amount']
+// exclude components with `variant` prop other than body
+const shouldBeExcluded = (attributes: JSXAttribute[]) =>
```
Comment (teimurjan) [inline packages/picasso-codemod/src/v17.0.0/typography-sizes/typography-sizes.ts:17]: I'd rename, so we can avoid the comment with the explanation.

```suggestion
const hasVariantBodyAttribute = (attributes: JSXAttribute[]) =>
```

File: packages/picasso-codemod/src/v17.0.0/typography-sizes/typography-sizes.ts [hunk 4]
```diff
@@ -31,22 +40,20 @@ const transform: Transform = (file, api, options) => {
...
-    .find(j.JSXAttribute, {
-      name: {
-        type: 'JSXIdentifier',
-        name: 'size'
+    .find(j.JSXAttribute, ({ name: { name }, start }) => {
+      if (name !== 'size' || transformedAttributes.includes(start)) {
+        return false
       }
+
+      transformedAttributes.push(start)
```
Comment (dmaklygin1) [inline packages/picasso-codemod/src/v17.0.0/typography-sizes/typography-sizes.ts:52]: I'm jus thinking, shouldn't we reset the `transformedAttributes` before switching to the next JSX node?

File: .changeset/thirty-bottles-complain.md (hunk 1 above)
Comment (OndrejTuma) [reply .changeset/thirty-bottles-complain.md]: Like it 👍

File: packages/picasso-codemod/src/v17.0.0/typography-sizes/typography-sizes.ts (hunk 2 above)
Comment (OndrejTuma) [reply packages/picasso-codemod/src/v17.0.0/typography-sizes/typography-sizes.ts]: good idea ✅  thanks

File: packages/picasso-codemod/src/v17.0.0/typography-sizes/typography-sizes.ts (hunk 3 above)
Comment (OndrejTuma) [reply packages/picasso-codemod/src/v17.0.0/typography-sizes/typography-sizes.ts]: I updated the function a little bit so I could use this naming which I like - it's `hasVariantHeadingAttribute`

File: packages/picasso-codemod/src/v17.0.0/typography-sizes/typography-sizes.ts (hunk 4 above)
Comment (OndrejTuma) [reply packages/picasso-codemod/src/v17.0.0/typography-sizes/typography-sizes.ts:52]: Actually this is important, it holds information of all changed attributes - `find()` method search for attributes even within children and that's why on nested Typography was codemod applied multiple times - and that caused the bug: `small --> xxsmall`

Comment (OndrejTuma) [reply packages/picasso-codemod/src/v17.0.0/typography-sizes/typography-sizes.ts]: OK so I changed my approach so its clearer now without the need to use `transformedAttributes`
cc @dmaklygin1

**Review by dmaklygin1 — APPROVED** [review]

**Review by teimurjan — APPROVED** [review]
👍

**Pull request merged**